agenthub/codeact_agent/codeact_agent.py

```diff
@@ -191,48 +210,71 @@ def step(self, state: State) -> Action:
         if latest_user_message and latest_user_message.strip() == '/exit':
             return AgentFinishAction()
 
-        # prepare what we want to send to the LLM
-        messages: list[dict[str, str]] = self._get_messages(state)
-
-        response = self.llm.completion(
-            messages=messages,
-            stop=[
-                '</execute_ipython>',
-                '</execute_bash>',
-                '</execute_browse>',
-            ],
-            temperature=0.0,
-        )
+        response = None
+        # give it multiple chances to get a response
+        # if it fails, we'll try to condense memory
+        attempt = 0
+        while not response and attempt < self.llm.config.attempts_to_condense:
+            # prepare what we want to send to the LLM
+            messages: list[Message] = self._get_messages(state)
+            print('No of tokens, ' + str(self.llm.get_token_count(messages)) + '\n')
+            response = self.llm.completion(
+                messages=messages,
+                stop=[
+                    '</execute_ipython>',
+                    '</execute_bash>',
+                    '</execute_browse>',
+                ],
+                temperature=0.0,
+                condense=True,
+            )
+            attempt += 1
+
         return self.action_parser.parse(response)
 
-    def _get_messages(self, state: State) -> list[dict[str, str]]:
+    def search_memory(self, query: str) -> list[str]:
+        raise NotImplementedError('Implement this abstract method')
+
+    def _get_messages(self, state: State) -> list[Message]:
         messages = [
-            {'role': 'system', 'content': self.system_message},
-            {'role': 'user', 'content': self.in_context_example},
+            Message(
+                message={'role': 'system', 'content': self.system_message},
+                condensable=False,
+            ),
+            Message(
+                message={'role': 'user', 'content': self.in_context_example},
+                condensable=False,
+            ),
         ]
 
         for event in state.history.get_events():
-            # create a regular message from an event
-            if isinstance(event, Action):
-                message = self.get_action_message(event)
-            elif isinstance(event, Observation):
-                message = self.get_observation_message(event)
-            else:
-                raise ValueError(f'Unknown event type: {type(event)}')
-
-            # add regular message
-            if message:
-                messages.append(message)
+            if event.id >= state.history.last_summarized_event_id:
+                if isinstance(event, AgentSummarizeAction):
```

Comment:
Aha! This is where AgentSummarizeAction is never found: it's because it was not saved anywhere, so it doesn't have an id. It doesn't need one, if it's managed by history/state like a part of the runtime state, because the `history` can pick it up (and does), but then we cannot do here `if event.id >=state.history.last_summarized_event_id` because the summarize action has the default id = -1.

Just checking isinstance outside this conditional should do.

Reply:
Thanks for pointing it out. Corrected it.